# CÁC THƯ VIỆN CẦN THIẾT

In [1]:
from datetime import datetime
import pandas as pd
import numpy as np

# I. INSPECT DATA PARQUET

In [2]:
day = pd.read_parquet("../../data/data_hitech_day.parquet")
month = pd.read_parquet("../../data/data_hitech_month.parquet")

In [3]:
day.head()

,cus_id,ten_phuongxa_hoatdong_chinh,ten_quanhuyen_hoatdong_chinh,ma_tinh_hoatdong_chinh,tongdon_diaban_chinh,tong_tien,min_tongtien,max_tongtien,tong_don,avg_tongtien,...,lydo_hoan,so_donhoan_nhieunhat,lydo_khieunai,so_khieunai_nhieunhat,ngay_hoptac,ngay_sinh,ten_dangky,nganh_hang,loai_khachhang,partition
0,"{'member0': 4110769.0, 'member1': None}",PHƯỜNG TÂN BIÊN,THÀNH PHỐ BIÊN HÒA,DNI,2,34500,16500,18000,2,17250.000000,...,None,NaN,NaN,NaN,28/06/2024,NaN,Pham******,Khác,Cá nhân,20251208
1,"{'member0': 4110774.0, 'member1': None}",PHƯỜNG LONG BÌNH,THÀNH PHỐ BIÊN HÒA,DNI,3,58000,15000,28000,3,19333.333984,...,None,NaN,NaN,NaN,28/11/2018,NaN,PHẠM******,Khác,Hộ kinh doanh cá thể,20251208
2,"{'member0': 4110777.0, 'member1': None}",PHƯỜNG QUANG TRUNG,QUẬN ĐỐNG ĐA,HNI,2,48000,17000,31000,2,24000.000000,...,None,NaN,NaN,NaN,08/01/2021,9.481536e+11,PHUNG******,Khác,Cá nhân,20251208
3,"{'member0': 4110779.0, 'member1': None}",None,HUYỆN AN DƯƠNG,HPG,1,27599,27599,27599,1,27599.000000,...,None,NaN,NaN,NaN,30/08/2018,NaN,******,Khác,Cá nhân,20251208
4,"{'member0': 4110850.0, 'member1': None}",PHƯỜNG QUAN HOA,QUẬN CẦU GIẤY,HNI,1,37000,37000,37000,1,37000.000000,...,None,NaN,NaN,NaN,30/06/2021,NaN,Tiên******,Khác,Cá nhân,20251208


In [4]:
print(day.dtypes)

cus_id                           object
ten_phuongxa_hoatdong_chinh      object
ten_quanhuyen_hoatdong_chinh     object
ma_tinh_hoatdong_chinh           object
tongdon_diaban_chinh              int32
tong_tien                         int32
min_tongtien                      int32
max_tongtien                      int32
tong_don                          int32
avg_tongtien                    float32
trong_luong                       int32
min_trongluong                    int32
max_trongluong                    int32
thu_ho                          float64
don_ptc                         float64
don_hoan                        float64
sanluong_phat_yeucau            float64
don_fail                        float64
khachhang_nhan                    int32
khachhang_nhieudon                int32
khachhang_moi_flag_daily          int32
sanluong_cptn                     int32
sanluong_exp                      int32
sanluong_fwd                      int32
sanluong_kho                      int32


In [5]:
month.head()

,cus_id,p5_tongtien,p10_tongtien,p25_tongtien,p50_tongtien,p75_tongtien,p90_tongtien,p95_tongtien,tongdon_cao,tongdon_thap,sanluong_phat_yeucau,khachhang_nhan,khachhang_nhieudon,khachhang_moi,khachhang_thangtruoc,khachhang_quaylai,khachhang_roibo,partition
0,"{'member0': 4110769.0, 'member1': None}",16500.000000,16500.000000,16500.000000,21000.0000,22000.000000,22000.0,24000.0,5,7,21.0,18,2,15.0,10.0,0.0,10.0,202511
1,"{'member0': 4110774.0, 'member1': None}",17335.238281,17999.000000,17999.000000,17999.0000,17999.000000,17999.0,17999.0,1773,1827,1943.0,1504,259,1174.0,1219.0,260.0,959.0,202511
2,"{'member0': 4110777.0, 'member1': None}",13000.000000,13000.000000,26799.777344,35550.1875,66610.789062,175211.0,229744.0,19,20,189.0,181,8,154.0,171.0,14.0,157.0,202511
3,"{'member0': 4110779.0, 'member1': None}",5000.000000,17333.333984,19000.000000,28000.0000,28000.000000,28000.0,28000.0,208,157,388.0,303,44,265.0,533.0,28.0,505.0,202511
4,"{'member0': 4110850.0, 'member1': None}",16500.000000,16500.000000,16500.000000,21501.0000,33002.000000,33002.0,33002.0,2,2,5.0,4,1,2.0,5.0,2.0,3.0,202511


In [6]:
print(month.dtypes)

cus_id                   object
p5_tongtien             float32
p10_tongtien            float32
p25_tongtien            float32
p50_tongtien            float32
p75_tongtien            float32
p90_tongtien            float32
p95_tongtien            float32
tongdon_cao               int32
tongdon_thap              int32
sanluong_phat_yeucau    float64
khachhang_nhan            int32
khachhang_nhieudon        int32
khachhang_moi           float64
khachhang_thangtruoc    float64
khachhang_quaylai       float64
khachhang_roibo         float64
partition                 int32
dtype: object


# II. THIẾT KẾ FEATURE TABLE

## 1. Xử lý `ngay_sinh` (dob)

**Thực trạng dữ liệu**
`ngay_sinh`:
- Missing rất nhiều
- Kiểu `float64`
- Có giá trị dạng epoch milliseconds. Ví dụ: 455155200000

Không thể bỏ cột này vì:
- Dùng để suy ra age merchant
- Age là biến quan trọng trong scoring

**Quy tắc xử lý**

| Trường hợp         | Cách xử lý              |
| ------------------ | ----------------------- |
| `ngay_sinh` = null | Fill = **avg tuổi**     |
| `ngay_sinh` hợp lệ | Chuẩn hoá → date → tuổi |

In [7]:
def normalize_dob(df):
    df = df.copy()

    # 1. ép về numeric để loại string / object / dict
    dob_raw = pd.to_numeric(df['ngay_sinh'], errors='coerce')

    # 2. loại bỏ inf / -inf (BẮT BUỘC)
    dob_raw = dob_raw.replace([np.inf, -np.inf], np.nan)

    # 3. giới hạn khoảng DOB hợp lý (1900-01-01 → 2010-01-01)
    min_ms = pd.Timestamp("1900-01-01").value // 10**6
    max_ms = pd.Timestamp("2010-01-01").value // 10**6

    # ⚠️ CLIP thay vì where → không overflow
    dob_raw = dob_raw.clip(lower=min_ms, upper=max_ms)

    # 4. convert sang datetime (AN TOÀN)
    df['dob'] = pd.to_datetime(
        dob_raw,
        unit='ms',
        errors='coerce'
    )

    # 5. tính tuổi
    today = pd.Timestamp.today()
    df['age_years'] = ((today - df['dob']).dt.days / 365.25)

    # 6. fill missing bằng tuổi trung bình
    avg_age = df['age_years'].mean()
    df['age_years'] = df['age_years'].fillna(avg_age)

    return df

day = normalize_dob(day)
day.head()

,cus_id,ten_phuongxa_hoatdong_chinh,ten_quanhuyen_hoatdong_chinh,ma_tinh_hoatdong_chinh,tongdon_diaban_chinh,tong_tien,min_tongtien,max_tongtien,tong_don,avg_tongtien,...,lydo_khieunai,so_khieunai_nhieunhat,ngay_hoptac,ngay_sinh,ten_dangky,nganh_hang,loai_khachhang,partition,dob,age_years
0,"{'member0': 4110769.0, 'member1': None}",PHƯỜNG TÂN BIÊN,THÀNH PHỐ BIÊN HÒA,DNI,2,34500,16500,18000,2,17250.000000,...,NaN,NaN,28/06/2024,NaN,Pham******,Khác,Cá nhân,20251208,NaT,30.650441
1,"{'member0': 4110774.0, 'member1': None}",PHƯỜNG LONG BÌNH,THÀNH PHỐ BIÊN HÒA,DNI,3,58000,15000,28000,3,19333.333984,...,NaN,NaN,28/11/2018,NaN,PHẠM******,Khác,Hộ kinh doanh cá thể,20251208,NaT,30.650441
2,"{'member0': 4110777.0, 'member1': None}",PHƯỜNG QUANG TRUNG,QUẬN ĐỐNG ĐA,HNI,2,48000,17000,31000,2,24000.000000,...,NaN,NaN,08/01/2021,9.481536e+11,PHUNG******,Khác,Cá nhân,20251208,2000-01-18,25.908282
3,"{'member0': 4110779.0, 'member1': None}",None,HUYỆN AN DƯƠNG,HPG,1,27599,27599,27599,1,27599.000000,...,NaN,NaN,30/08/2018,NaN,******,Khác,Cá nhân,20251208,NaT,30.650441
4,"{'member0': 4110850.0, 'member1': None}",PHƯỜNG QUAN HOA,QUẬN CẦU GIẤY,HNI,1,37000,37000,37000,1,37000.000000,...,NaN,NaN,30/06/2021,NaN,Tiên******,Khác,Cá nhân,20251208,NaT,30.650441


## 2. Chuẩn hoá khoá định danh merchant

Trong dữ liệu gốc:
- cus_id là một object dạng dict
- ID thực tế nằm ở cus_id['member0']

→ Ta chuẩn hoá thành merchant_id để dùng xuyên suốt.

In [8]:
def extract_merchant_id(x):
    if isinstance(x, dict):
        return x.get("member0")
    return None

day["merchant_id"] = day["cus_id"].apply(extract_merchant_id)
month["merchant_id"] = month["cus_id"].apply(extract_merchant_id)

## 3. Join dữ liệu day & month

**Logic**
- Base table: daily snapshot
- Monthly data: bổ sung hành vi dài hạn
- Join: LEFT JOIN

In [9]:
df = day.merge(
    month[
        [
            "merchant_id",
            "tongdon_cao",
            "tongdon_thap",
            "khachhang_quaylai",
            "khachhang_roibo"
        ]
    ],
    on="merchant_id",
    how="left"
)

df.head()

,cus_id,ten_phuongxa_hoatdong_chinh,ten_quanhuyen_hoatdong_chinh,ma_tinh_hoatdong_chinh,tongdon_diaban_chinh,tong_tien,min_tongtien,max_tongtien,tong_don,avg_tongtien,...,nganh_hang,loai_khachhang,partition,dob,age_years,merchant_id,tongdon_cao,tongdon_thap,khachhang_quaylai,khachhang_roibo
0,"{'member0': 4110769.0, 'member1': None}",PHƯỜNG TÂN BIÊN,THÀNH PHỐ BIÊN HÒA,DNI,2,34500,16500,18000,2,17250.000000,...,Khác,Cá nhân,20251208,NaT,30.650441,4110769.0,5,7,0.0,10.0
1,"{'member0': 4110774.0, 'member1': None}",PHƯỜNG LONG BÌNH,THÀNH PHỐ BIÊN HÒA,DNI,3,58000,15000,28000,3,19333.333984,...,Khác,Hộ kinh doanh cá thể,20251208,NaT,30.650441,4110774.0,1773,1827,260.0,959.0
2,"{'member0': 4110777.0, 'member1': None}",PHƯỜNG QUANG TRUNG,QUẬN ĐỐNG ĐA,HNI,2,48000,17000,31000,2,24000.000000,...,Khác,Cá nhân,20251208,2000-01-18,25.908282,4110777.0,19,20,14.0,157.0
3,"{'member0': 4110779.0, 'member1': None}",None,HUYỆN AN DƯƠNG,HPG,1,27599,27599,27599,1,27599.000000,...,Khác,Cá nhân,20251208,NaT,30.650441,4110779.0,208,157,28.0,505.0
4,"{'member0': 4110850.0, 'member1': None}",PHƯỜNG QUAN HOA,QUẬN CẦU GIẤY,HNI,1,37000,37000,37000,1,37000.000000,...,Khác,Cá nhân,20251208,NaT,30.650441,4110850.0,2,2,2.0,3.0


## 4. Xử lý missing sau join

Monthly data có thể thiếu (merchant mới, inactive…)

**Quy ước Phase 1**
- Thiếu = 0 (chưa phát sinh hành vi)

In [10]:
month_cols = [
    "tongdon_cao",
    "tongdon_thap",
    "khachhang_quaylai",
    "khachhang_roibo"
]

df[month_cols] = df[month_cols].fillna(0)

## 5. Thời gian hợp tác (Usage Duration)

Ý nghĩa
- Merchant hợp tác lâu → ổn định hơn
- Feature quan trọng cho credit scoring

In [11]:
df["ngay_hoptac_dt"] = pd.to_datetime(
    df["ngay_hoptac"],
    dayfirst=True,
    errors="coerce"
)

df["usage_duration_months"] = (
    (datetime.today() - df["ngay_hoptac_dt"]).dt.days / 30
)

df["usage_duration_months"].describe()

count    10336.000000
mean        41.355541
std         29.455970
min          0.233333
25%         14.133333
50%         38.866667
75%         64.233333
max        100.000000
Name: usage_duration_months, dtype: float64

## 6. Làm sạch các chỉ số giao dịch chính

Áp dụng cho Phase 1:
- Missing → 0
- Không scale / normalize (để rule-based dễ hiểu)

In [12]:
numeric_cols = [
    "tong_tien",
    "tong_don",
    "avg_tongtien"
]

df[numeric_cols] = df[numeric_cols].fillna(0)

## 7. Feature table

In [13]:
selected_cols = [
    "merchant_id",
    "ma_tinh_hoatdong_chinh",
    "tong_tien",
    "tong_don",
    "avg_tongtien",
    "tongdon_cao",
    "tongdon_thap",
    "khachhang_quaylai",
    "khachhang_roibo",
    "age_years",
    "usage_duration_months",
    "partition"
]

df[selected_cols].head()

,merchant_id,ma_tinh_hoatdong_chinh,tong_tien,tong_don,avg_tongtien,tongdon_cao,tongdon_thap,khachhang_quaylai,khachhang_roibo,age_years,usage_duration_months,partition
0,4110769.0,DNI,34500,2,17250.000000,5,7,0.0,10.0,30.650441,17.833333,20251208
1,4110774.0,DNI,58000,3,19333.333984,1773,1827,260.0,959.0,30.650441,85.800000,20251208
2,4110777.0,HNI,48000,2,24000.000000,19,20,14.0,157.0,25.908282,60.066667,20251208
3,4110779.0,HPG,27599,1,27599.000000,208,157,28.0,505.0,30.650441,88.800000,20251208
4,4110850.0,HNI,37000,1,37000.000000,2,2,2.0,3.0,30.650441,54.300000,20251208


## 8. Kiểm tra chất lượng dữ liệu

In [14]:
df.shape, df["merchant_id"].nunique()

((10342, 46), 9981)

**Ý nghĩa con số**
- Số dòng: 10,342
- Số merchant_id unique: 9,981
→ Có 361 dòng bị trùng merchant_id

**Nguyên nhân có thể xảy ra**
- `df_day` không đảm bảo 1 merchant = 1 dòng. Có thể do:
    - Nhiều snapshot theo ngày
    - Nhiều partition
    - Nhiều bản ghi do data upstream
    - Monthly data KHÔNG gây duplication, vì join là LEFT từ day.
- `merchant_id` có null


In [15]:
# Số merchant bị lặp
(df["merchant_id"].value_counts() > 1).sum()

np.int64(0)

In [16]:
# Số merchant có merchant_id = null
df["merchant_id"].isna().sum()

np.int64(361)

In [17]:
# Drop
df = df[df["merchant_id"].notna()]
df.shape, df["merchant_id"].nunique()

((9981, 46), 9981)

# III. RULE-BASED ELIGIBILITY (THEO CHUẨN VTP)

Áp dụng đúng policy VTP để xác định:
- Merchant ĐỦ ĐIỀU KIỆN hay KHÔNG
- Nếu không đạt → fail ở tiêu chí nào

**Đặc điểm**
- Hard rule (Boolean)
- Không cộng điểm
- Không bù trừ
- Fail 1 điều kiện = fail toàn bộ

## 1. Tổng hợp rule VTP (official)

| STT | Tiêu chí                           | Điều kiện                |
| --- | ---------------------------------- | ------------------------ |
| 1   | Tuổi                               | 20 ≤ age ≤ 50            |
| 2   | Thời gian sử dụng                  | ≥ 6 tháng                |
| 3   | Địa bàn                            | ≥ 48 tỉnh                |
| 4   | Doanh thu TB/tháng (6 tháng)       | ≥ 10 triệu               |
| 5   | Không có tháng < 3 triệu (6 tháng) | min ≥ 3 triệu            |
| 6   | Số đơn TB/tháng (6 tháng)          | ≥ 10                     |
| 7   | Doanh thu 3 quý                    | Không giảm >30% liên tục |
| 8   | Doanh thu 4 tháng                  | Không giảm >20% liên tục |

**LƯU Ý RẤT QUAN TRỌNG VỀ DATA**

Dataset hiện tại:
- **KHÔNG** có time series theo tháng
- **KHÔNG** có province count
- **KHÔNG** có quarterly revenue

Vì vậy:
- Các rule **CÓ** data → triển khai trực tiếp
- Các rule **CHƯA** có data → đánh dấu NOT_EVALUATED

→ Đây là cách làm đúng trong thực tế, không được tự suy diễn. Sẽ bổ sung trong phase 2.

## 2. Khởi tạo bảng rule check

In [18]:
rule_df = df.copy()

## 3. RULE 1 – Tuổi (20–50)

In [19]:
rule_df["rule_age"] = (
    (rule_df["age_years"] >= 20)
    & (rule_df["age_years"] <= 50)
)

## 4. RULE 2 – Thời gian sử dụng ≥ 6 tháng

In [20]:
rule_df["rule_usage_duration"] = (
    rule_df["usage_duration_months"] >= 6
)

## 5. RULE 3 – Địa bàn thuộc 48 tỉnh (CHƯA THỂ ĐỐI CHIẾU)

Hiện tại:
- Chỉ có `ma_tinh_hoatdong_chinh`

→ Rule này sẽ được triển khai khi có bảng valid_provinces.

## 6. RULE 4 – Doanh thu TB ≥ 10tr/tháng (6 tháng)

Yêu cầu:
```bash
    sum(revenue_last_6m) / 6 ≥ 10,000,000
```

Hiện tại:
- `tong_tien` = lifetime
- Không có window 6 tháng

In [21]:
rule_df["rule_avg_monthly_revenue_6m"] = None

## 7. RULE 5 – Không có tháng < 3tr (6 tháng)

Yêu cầu:
```bash
    min(monthly_revenue_last_6m) ≥ 3,000,000
```

In [22]:
rule_df["rule_min_monthly_revenue_6m"] = None

## 8. RULE 6 – Số đơn TB/tháng ≥ 10 (6 tháng)

In [23]:
rule_df["rule_avg_monthly_orders_6m"] = None

## 10. RULE 8 – Doanh thu 4 tháng không giảm >20%

In [24]:
rule_df["rule_monthly_decline_4m"] = None

## 11. Tổng hợp kết quả rule

### a. Xác định rule CÓ THỂ ĐÁNH GIÁ

In [25]:
evaluated_rules = [
    "rule_age",
    "rule_usage_duration"
]

### b. Merchant đạt tất cả rule đã đánh giá?

In [26]:
rule_df["eligible_partial"] = rule_df[evaluated_rules].all(axis=1)
rule_df.head()

,cus_id,ten_phuongxa_hoatdong_chinh,ten_quanhuyen_hoatdong_chinh,ma_tinh_hoatdong_chinh,tongdon_diaban_chinh,tong_tien,min_tongtien,max_tongtien,tong_don,avg_tongtien,...,khachhang_roibo,ngay_hoptac_dt,usage_duration_months,rule_age,rule_usage_duration,rule_avg_monthly_revenue_6m,rule_min_monthly_revenue_6m,rule_avg_monthly_orders_6m,rule_monthly_decline_4m,eligible_partial
0,"{'member0': 4110769.0, 'member1': None}",PHƯỜNG TÂN BIÊN,THÀNH PHỐ BIÊN HÒA,DNI,2,34500,16500,18000,2,17250.000000,...,10.0,2024-06-28,17.833333,True,True,None,None,None,None,True
1,"{'member0': 4110774.0, 'member1': None}",PHƯỜNG LONG BÌNH,THÀNH PHỐ BIÊN HÒA,DNI,3,58000,15000,28000,3,19333.333984,...,959.0,2018-11-28,85.800000,True,True,None,None,None,None,True
2,"{'member0': 4110777.0, 'member1': None}",PHƯỜNG QUANG TRUNG,QUẬN ĐỐNG ĐA,HNI,2,48000,17000,31000,2,24000.000000,...,157.0,2021-01-08,60.066667,True,True,None,None,None,None,True
3,"{'member0': 4110779.0, 'member1': None}",None,HUYỆN AN DƯƠNG,HPG,1,27599,27599,27599,1,27599.000000,...,505.0,2018-08-30,88.800000,True,True,None,None,None,None,True
4,"{'member0': 4110850.0, 'member1': None}",PHƯỜNG QUAN HOA,QUẬN CẦU GIẤY,HNI,1,37000,37000,37000,1,37000.000000,...,3.0,2021-06-30,54.300000,True,True,None,None,None,None,True


## 12. Ghi rõ lý do FAIL

In [27]:
def fail_reasons(row):
    reasons = []

    if not row["rule_age"]:
        reasons.append("Tuổi không trong khoảng 20–50")

    if not row["rule_usage_duration"]:
        reasons.append("Thời gian sử dụng < 6 tháng")

    return "; ".join(reasons)

rule_df["fail_reason"] = rule_df.apply(fail_reasons, axis=1)

### 13. Kết quả eligibility (Phase 1 – theo data hiện có)

In [28]:
rule_df[
    [
        "merchant_id",
        "eligible_partial",
        "fail_reason"
    ]
].head(10)

,merchant_id,eligible_partial,fail_reason
0,4110769.0,True,
1,4110774.0,True,
2,4110777.0,True,
3,4110779.0,True,
4,4110850.0,True,
5,4110859.0,True,
6,4110864.0,True,
7,4110866.0,True,
8,4110867.0,True,
9,4110882.0,True,


## 14. Chuẩn bị output (MINIMUM VIABLE OUTPUT)

In [29]:
eligibility_output = rule_df[
    [
        "merchant_id",
        "eligible_partial",
        "fail_reason",
        "partition"
    ]
]

eligibility_output.to_excel(
    "../output/scoring_output.xlsx",
    index=False
)

# IV. PHÂN TÍCH KHOẢNG TRỐNG DỮ LIỆU & BẢNG CẦN BỔ SUNG

## 1. Tổng hợp trạng thái các tiêu chí VTP

| STT | Tiêu chí                            | Trạng thái      | Nguyên nhân                  |
| --- | ----------------------------------- | --------------- | ---------------------------- |
| 1   | Tuổi (20–50)                        | ✅ Đã đánh giá   | Có DOB                       |
| 2   | Thời gian sử dụng ≥ 6 tháng         | ✅ Đã đánh giá   | Có ngày hợp tác              |
| 3   | Địa bàn ≥ 48 tỉnh                   | ❌ Chưa đánh giá | Thiếu dữ liệu theo tỉnh      |
| 4   | Doanh thu TB ≥ 10tr/tháng (6 tháng) | ❌ Chưa đánh giá | Không có dữ liệu theo tháng  |
| 5   | Không có tháng < 3tr (6 tháng)      | ❌ Chưa đánh giá | Không có min doanh thu tháng |
| 6   | Số đơn TB ≥ 10/tháng (6 tháng)      | ❌ Chưa đánh giá | Không có số đơn theo tháng   |
| 7   | Doanh thu 3 quý không giảm >30%     | ❌ Chưa đánh giá | Không có dữ liệu theo quý    |
| 8   | Doanh thu 4 tháng không giảm >20%   | ❌ Chưa đánh giá | Không có time series tháng   |

## 2. Phân loại khoảng trống dữ liệu

**Nhóm A – Thiếu dữ liệu theo thời gian**
(Critical)

Các rule cần:
- Doanh thu theo tháng
- Số đơn theo tháng
- Gom nhóm theo quý
→ Dataset hiện tại chỉ là snapshot, không có lịch sử.

**Nhóm B – Thiếu dữ liệu theo địa bàn**
(Critical)

Rule yêu cầu:
- Merchant hoạt động tại ≥ 48 tỉnh
Hiện tại:
 - Chỉ có ma_tinh_hoatdong_chinh.
→ Cần bổ sung danh sách mã tỉnh valid

## 3. BẢNG DỮ LIỆU CẦN BỔ SUNG (ĐỀ XUẤT)

**Bảng 1 – merchant_monthly_summary**
Mục đích:
- Đánh giá các rule 4, 5, 6, 8
Grain:
- 1 merchant – 1 tháng

| Cột           | Ý nghĩa                   |
| ------------- | ------------------------- |
| merchant_id   | Mã khách hàng             |
| year_month    | Tháng (YYYYMM)            |
| total_revenue | Tổng doanh thu thành công |
| total_orders  | Tổng đơn thành công       |
| snapshot_date | Ngày chốt                 |

Yêu cầu:
- Ít nhất 6 tháng liên tục
- Không được thiếu tháng ở giữa

**Bảng 2 – merchant_quarterly_revenue**
Mục đích
- Đánh giá rule 7 (giảm doanh thu theo quý)
Grain
- 1 merchant – 1 quý

| Cột           | Ý nghĩa          |
| ------------- | ---------------- |
| merchant_id   | Mã khách hàng    |
| year_quarter  | Quý (VD: 2024Q3) |
| total_revenue | Doanh thu quý    |

Có thể derive từ bảng tháng

**Bảng 3 - valid_provinces**
Mục đích:
- Làm reference table cho rule địa bàn.
Grain:
- 1 dòng = 1 tỉnh hợp lệ

| Cột            | Ý nghĩa                    |
| -------------- | -------------------------- |
| province_code  | Mã tỉnh                    |
| province_name  | Tên tỉnh                   |
| is_active      | Có đang áp dụng hay không  |

Ví dụ:

| province_code | province_name | is_active |
| ------------- | ------------- | --------- |
| HNI           | Hà Nội        | 1         |
| HCM           | TP.HCM        | 1         |
| DNI           | Đồng Nai      | 1         |
| …             | …             | …         |


## 4. Mapping bảng → rule

| Rule   | Bảng cần                   |
| ------ | -------------------------- |
| Rule 3 | merchant_province_activity |
| Rule 4 | merchant_monthly_summary   |
| Rule 5 | merchant_monthly_summary   |
| Rule 6 | merchant_monthly_summary   |
| Rule 7 | merchant_quarterly_revenue |
| Rule 8 | merchant_monthly_summary   |


## 5. Thứ tự ưu tiên triển khai (khuyến nghị)

1. `merchant_monthly_summary`
→ mở được 5/8 rule

2. `merchant_province_activity`
→ mở rule địa bàn

3. `merchant_quarterly_revenue`
→ derive từ bảng tháng

Chỉ cần bảng tháng là đã unlock gần hết policy

## 6. Phiên bản Phase 1 đề xuất

| Phiên bản | Phạm vi          |
| --------- | ---------------- |
| Phase 1.0 | Rule 1–2         |
| Phase 1.1 | Rule 1–2–4–5–6–8 |
| Phase 1.2 | Rule 1–8 đầy đủ  |

## 7. Thông điệp
- Hệ thống hiện đã sẵn sàng về mặt logic và kiến trúc.
- Việc chưa áp dụng đầy đủ rule là do hạn chế dữ liệu lịch sử, không phải do mô hình.
- Khi bổ sung bảng tháng, toàn bộ policy sẽ được kích hoạt ngay.

# V. ĐỀ XUẤT KIẾN TRÚC PIPELINE KHI CÓ ĐỦ DỮ LIỆU

## 1. Nguyên tắc thiết kế pipeline
1. Tách rõ tầng dữ liệu
2. Raw → Aggregation → Rule
3. Mỗi bảng có grain rõ ràng
4. Rule không xử lý raw
5. Có thể tái chạy (re-run)
6. Trace được lý do đạt / không đạt

## 2. Tổng quan kiến trúc pipeline

## 3. Chi tiết từng tầng dữ liệu

### a. RAW DATA ZONE

Nguồn
- Đơn hàng thành công
- Địa bàn phát sinh đơn
- Thời gian giao dịch

Grain
- 1 dòng = 1 đơn hàng

Yêu cầu tối thiểu
- merchant_id
- order_date
- province_code
- order_value
- order_status

**Không xử lý nghiệp vụ ở tầng này.**

### b. AGGREGATION ZONE

**Bảng 1 – merchant_monthly_summary**

| Thuộc tính    | Mô tả          |
| ------------- | -------------- |
| merchant_id   | Mã KH          |
| year_month    | Tháng          |
| total_revenue | Tổng doanh thu |
| total_orders  | Tổng đơn       |

→ Phục vụ rule 4, 5, 6, 8

**Bảng 2 – merchant_quarterly_revenue**

| Thuộc tính    | Mô tả         |
| ------------- | ------------- |
| merchant_id   | Mã KH         |
| year_quarter  | Quý           |
| total_revenue | Doanh thu quý |

→ Phục vụ rule 7

Bảng 3 – merchant_province_activity
| Thuộc tính    | Mô tả    |
| ------------- | -------- |
| merchant_id   | Mã KH    |
| province_code | Tỉnh     |
| total_orders  | Tổng đơn |

→ Phục vụ rule 3

**→ Các bảng này tự động cập nhật theo lịch**

### c. RULE ENGINE

Input
- Feature merchant-level
- Bảng aggregation
- Bảng tỉnh hợp lệ

Xử lý
- Áp dụng từng rule độc lập
- Ghi kết quả từng rule (True/False)
- Tổng hợp eligibility

Output trung gian
- rule_evaluation_table

| merchant_id | rule_1 | rule_2 | … | rule_8 |
| ----------- | ------ | ------ | - | ------ |
|             |        |        | … |        |

### d. OUTPUT ZONE

Bảng cuối
`scoring_output`

| merchant_id | eligible | fail_reason | snapshot_date |
| ----------- | -------- | ----------- | ------------- |
|             |          |             |               |

**Đây là bảng duy nhất được cấp cho nghiệp vụ.**

## 4. Lịch chạy đề xuất

| Pipeline              | Tần suất   |
| --------------------- | ---------- |
| Raw ingest            | Hàng ngày  |
| Monthly aggregation   | Hàng tháng |
| Quarterly aggregation | Hàng quý   |
| Rule engine           | Hàng tháng |

**Rule engine luôn chạy sau khi aggregation hoàn tất.**

## 5. Quản lý version & thay đổi rule

Nguyên tắc
- Mỗi lần đổi rule → tạo version mới
- Không sửa ngược dữ liệu cũ

Ví dụ
- rule_version = VTP_2025_01
- rule_version = VTP_2025_06

## 6. Khả năng mở rộng (Phase 2)

Khi đã có pipeline ổn định:
- Thêm credit score (ranking)
- Thêm ML model
- Champion / challenger
- Theo dõi tỷ lệ default

**Nhưng rule VTP vẫn là điều kiện bắt buộc trước.**

## 7. Kiểm soát chất lượng dữ liệu

Bắt buộc có:
- Check thiếu tháng
- Check doanh thu âm
- Check merchant trùng
- Log lỗi pipeline